### PR TITLE
controllers: trigger reconcile event to storage client if hash is not matched

### PR DESCRIPTION
**Pull request by rchikatw**
- Save hash to storage client status
- add an annotation for the subscription channel to the storage client
- Add a new field to the status of the storage client
- Testing is pending will add results once it's done
- I have messed PR (https://github.com/red-hat-storage/ocs-client-operator/pull/200) which has similar changes and due to which its closed automatically hence creating the current pr.

### CODE REVIEW

File: service/status-report/main.go
Comment (rchikatw) [inline service/status-report/main.go:122]: Not using `CreateOrUpdate` because utils.AddAnnotations works what we need in the current situation.

File: internal/controller/storageclient_controller.go
Comment (leelavg) [inline internal/controller/storageclient_controller.go:218]: shouldn't this preferably be at the end of reconcile? In current logic and updated hash isn't suggesting a successful reconcile, isn't it?

Comment (rchikatw) [reply internal/controller/storageclient_controller.go:218]: I can move it to end not a problem

File: service/status-report/main.go
Comment (rchikatw) [reply service/status-report/main.go:132]: We have decided we won’t be touching subscriptions part ok. And if the hash changes we want to trigger the reconcile to storage client. I am patching annotation for config hash same way how it’s done for subscription.

File: internal/controller/storageclient_controller.go
Comment (rchikatw) [reply internal/controller/storageclient_controller.go]: Done.

**Comment by leelavg** [comment]
LGTM but will wait for Ohad's approval.

**Comment by nb-ohad** [comment]
/hold 

@rchikatw Something here seems to be in misalignment with the design. 
I can see 3 different values taking into consideration
1. The annotation
2. The value on the status
3. The value coming from the provider

One of them is unnecessary and redundant, I will vote out the status value.

**Comment by rchikatw** [comment]
> /hold
> 
> @rchikatw Something here seems to be in misalignment with the design. I can see 3 different values taking into consideration
> 
> 1. The annotation
> 2. The value on the status
> 3. The value coming from the provider
> 
> One of them is unnecessary and redundant, I will vote out the status value.

If we decided to remove the status section from storage client then there is no point in keeping the root field at the storage config response [here](https://github.com/red-hat-storage/ocs-operator/blob/main/services/provider/proto/provider.proto#L87)

File: pkg/utils/k8sutils.go
Comment (nb-ohad) [inline pkg/utils/k8sutils.go:45]: 1. This is the name of an annotation on the CR so by default it represents the desired state. There is no reason to mention desired in the name of the annotation. 
2. The encoding of the data `hash` might change in the future, I don't think we need to mention that as part of the name as it is an impl detail.
3. The annotation name does not describe what kind of function this annotation represents

File: service/status-report/main.go
Comment (nb-ohad) [inline service/status-report/main.go:125]: Are we sure that patch is the correct approach for this new annotation? What happens if a reconcile is in motion while you are trying to read and update the annotation. It might be that be the time you get to the patch command the CR already changed

**Review by nb-ohad — CHANGES_REQUESTED** [review]

**Comment by nb-ohad** [comment]
@rchikatw Why do the changes in this PR require changes in hundreds of vendor files?

File: service/status-report/main.go
Comment (rchikatw) [reply service/status-report/main.go:125]: I'm not sure which method I should use in this case. If I use "update" on CR, it sends the entire object to the API server, while "patch" sends only the changes that need to be made. There is a greater risk of conflict when using "update" and a lesser risk when using "patch," but in both cases, we can still encounter issues.
Even if I use PartialMetadata there will be same issue I feel

File: pkg/utils/k8sutils.go
Comment (rchikatw) [reply pkg/utils/k8sutils.go:45]: how about changing it to `consumer.config.state`

**Comment by leelavg** [comment]
> @rchikatw Why do the changes in this PR require changes in hundreds of vendor files?  
- should be solved by https://github.com/red-hat-storage/ocs-operator/pull/2798

**Comment by rchikatw** [comment]
> > @rchikatw Why do the changes in this PR require changes in hundreds of vendor files?
> 
> * should be solved by [provider: split provider related pkgs into 3 go modules  ocs-operator#2798](https://github.com/red-hat-storage/ocs-operator/pull/2798)

Sure once this is merged I will rebase it.

File: service/status-report/main.go
Comment (nb-ohad) [reply service/status-report/main.go:126]: We want the conflict because we do not want to update based on stale data. Let's use update instead of paths for this new annotation. The old annotation should continue to use patch

File: pkg/utils/k8sutils.go
Comment (nb-ohad) [reply pkg/utils/k8sutils.go:45]: What about `ocs.openshift.io/provider-side-state`? or something like that?

Comment (rchikatw) [reply pkg/utils/k8sutils.go:45]: just `ocs.openshift.io/provider-state` is good enough i guess

Comment (nb-ohad) [reply pkg/utils/k8sutils.go:45]: `provider-state` implies this represents the entire provider state which is wrong. 
`provider-side-state` implies this represent some state on the provider side

File: service/status-report/main.go
Comment (rchikatw) [reply service/status-report/main.go:126]: I will use `.Update` because it will consider the resource version while updating the resource.

**Comment by leelavg** [comment]
@rchikatw new proto changes aren't being pulled?

**Comment by rchikatw** [comment]
> @rchikatw new proto changes aren't being pulled?

It's pulled now. It took some time to push that change; meanwhile, you saw my PR.

**Review by leelavg — APPROVED** [review]

**Comment by leelavg** [comment]
/unhold

From Ohad
```
The PR is so big I am unable to access the git without my tab crashing

@lgangava Can you please remove the hold on my behalf?
```

**Pull request merged**
